### DISCUSSION TITLE
"Cannot mix incompatible Qt library with this library" & weird xdg-open behaviour

### DISCUSSION THREAD

**Discussion by AlexBocken** [qutebrowser/qutebrowser#6447] [Q&A] [answered]
**Version info**:
qutebrowser v2.2.1
Git commit: 413ce38ab on master (2021-05-05 20:36:27 +0200)
Backend: QtWebEngine 5.15.3, based on Chromium 87.0.4280.144
Qt: 5.15.2

CPython: 3.9.4
PyQt: 5.15.4

sip: 5.4.0
colorama: 0.4.4
jinja2: 2.11.3
pygments: 2.9.0
yaml: 5.4.1
adblock: 0.4.4
PyQt5.QtWebEngineWidgets: yes
PyQt5.QtWebEngine: 5.15.4
PyQt5.QtWebKitWidgets: yes
pdf.js: no
sqlite: 3.35.5
QtNetwork SSL: OpenSSL 1.1.1k  25 Mar 2021

Style: QFusionStyle
Platform plugin: xcb
OpenGL: Intel, 4.6 (Compatibility Profile) Mesa 21.0.3
Platform: Linux-5.11.16-arch1-1-x86_64-with-glibc2.33, 64bit
Linux distribution: Arch Linux (arch)
Frozen: False
Imported from /usr/lib/python3.9/site-packages/qutebrowser
Using Python from /usr/bin/python3
Qt library executable path: /usr/lib/qt/libexec, data path: /usr/share/qt

Paths:
cache: /home/alex/.cache/qutebrowser
config: /home/alex/.config/qutebrowser
data: /home/alex/.local/share/qutebrowser
runtime: /run/user/1002/qutebrowser
system data: /usr/share/qutebrowser

Autoconfig loaded: yes
Config.py: /home/alex/.config/qutebrowser/config.py has been loaded
Uptime: 0:00:36

**Does the bug happen if you start with `--temp-basedir`?**: Yes

**Description**

I've noticed that clicking on links in Anki, if qutebrowser is not already running, that qutebrowser crashes with the following backtrace:

```
15:35:16 DEBUG    init       app:_init_modules:453 Initializing logging from config...
15:35:16 DEBUG    init       log:init_from_config:543 --debug flag overrides log configs
15:35:16 DEBUG    init       app:_init_modules:457 Initializing save manager...
15:35:16 DEBUG    init       app:_init_modules:463 Checking backend requirements...
15:35:16 DEBUG    init       app:_init_modules:466 Initializing prompts...
15:35:16 DEBUG    init       app:_init_modules:469 Initializing network...
15:35:16 DEBUG    init       app:_init_modules:472 Initializing proxy...
15:35:16 DEBUG    init       app:_init_modules:476 Initializing downloads...
15:35:16 DEBUG    init       app:_init_modules:482 Initializing SQL...
15:35:16 CRITICAL qt         Unknown module:none:0 Cannot mix incompatible Qt library (5.15.1) with this library (5.15.2)
  File "/usr/bin/qutebrowser", line 33, in <module>
    sys.exit(load_entry_point('qutebrowser==2.2.1', 'gui_scripts', 'qutebrowser')())
  File "/usr/lib/python3.9/site-packages/qutebrowser/qutebrowser.py", line 220, in main
    return app.run(args)
  File "/usr/lib/python3.9/site-packages/qutebrowser/app.py", line 122, in run
    init(args=args)
  File "/usr/lib/python3.9/site-packages/qutebrowser/app.py", line 157, in init
    _init_modules(args=args)
  File "/usr/lib/python3.9/site-packages/qutebrowser/app.py", line 483, in _init_modules
    sql.init(os.path.join(standarddir.data(), 'history.sqlite'))
  File "/usr/lib/python3.9/site-packages/qutebrowser/misc/sql.py", line 173, in init
    database = QSqlDatabase.addDatabase('QSQLITE')
  File "/usr/lib/python3.9/site-packages/qutebrowser/utils/log.py", line 508, in qt_message_handler
    stack: Optional[str] = ''.join(traceback.format_stack())
```

This issue does not arise when using `xdg-open <link>` from a terminal. It also doesn't happen if qutebrowser is already running. I don't quite understand how Anki invokes xdg-open and how it's even possible to differ but apparently it does. It could also be a problem that Anki also relies on QtWebEngine? I don't believe this to be an Anki issue since other browser work flawlessly with Anki.

**How to reproduce**
As this seems quite specific to Anki for me (but I don't believe it to be an Anki bug?) the only reccomendation from me is to try to open a link in an Anki Card while xdg-open linking to qutebrowser.
I'm using a small wrapper script that get's initialized with browser.desktop.

I've seen https://github.com/qutebrowser/qutebrowser/issues/3167 which seems to have a similar issue. I believe I tried to fix this by reinstalling all qutebrowser dependencies via 
```
yay -S $(pactree -u qutebrowser-git)
``` 
but I'm not sure if that's enough. It didn't change anything for me. 

This might be an easy issue to fix but I don't really know how to fix it. Any hints?

**Comment by The-Compiler** [comment]
I'm converting this to a discussion as it seems like an issue with your system, not a bug in qutebrowser.

Usually this happens when you install a Qt plugin from the AUR, and then don't rebuild it when upgrading Qt. If it's not that, make sure your system is fully upgraded (partial upgrades are unsupported on Arch). FWIW I can't reproduce.

Could you please show the output of `find /usr/lib/qt -name '*.so' -exec pacman -Qo {} +`?
Also, could you show the output of running the whole thing (including Anki, ideally) with `export QT_DEBUG_PLUGINS=1` as well?

**Reply by AlexBocken** [reply]
Thanks for the quick reply. Apologies for creating an issue. I was not quite sure whether this should be a discussion or an issue, seems like I was in the wrong.

Here's my output from your given `find` command:
[qt_plugins.txt](https://github.com/qutebrowser/qutebrowser/files/6435276/qt_plugins.txt)

And with `export QT_DEBUG_PLUGINS=1` and starting anki without any addons via `anki --safemode` this is my log:
[crash_from_anki_log.txt](https://github.com/qutebrowser/qutebrowser/files/6435254/crash_from_anki_log.txt)

**Comment by The-Compiler** [comment, marked as answer]
Based on:

```
Qt debug: QFactoryLoader::QFactoryLoader() checking directory path "/usr/local/share/anki/bin/PyQt5/Qt/plugins/platforms" ...
```

it looks like you didn't install anki via the Archlinux packages, but instead installed some custom build of anki system-wide, which includes a custom build of PyQt5/Qt.

I'm guessing the second Qt build you have there is incompatible with the system-wide one, and launching qutebrowser from Anki gives you a weird mixture of the two, which is why Qt complains.

**Reply by AlexBocken** [reply]
Thank you so much for your help! I truly appreciate it. I will be able to fix it based on this information, I believe. I'm sorry for having used your time when it turns out this had pretty much nothing to do with qutebrowser.

edit: yes, switching from `anki-official-binary-bundle` to `anki-git` fixed the issue. (`community/anki` is too outdated for me).
